### PR TITLE
Feature attrdict and model definition declutter

### PR DESCRIPTION
**Pull request by irm-codebase**
Prepares our IO structure for `pydantic` by moving as much YAML read features as possible out of `AttrDict`.

## Summary of changes in this pull request

* YAML reading functionality has been moved out of `AttrDict`
* `templates:` improvements
    * now resolved before model definition reaches `calliope.Model`
    * `template:` can be used anywhere within a YAML file, not just `nodes`, `techs` or `data_tables`
    * cyclic template inheritance errors (template1 -> template2 -> template1 -> template2...) are now detected by using a stack
* removed `inheritance` helper function. It was made redundant by the template solving.
* Grouped `scenario`, `override_dict` and `templates` into `preprocess/model_definition.py` to ensure a single point of entry for defining their order of priority.

## Reviewer checklist

- [ ] Test(s) added to cover contribution
- [ ] Documentation updated
- [ ] Changelog updated
- [ ] Coverage maintained or improved

### CODE REVIEW

**Comment by irm-codebase** [comment]
Some additional context @brynpickering:
One of the main ideas behind this PR is to make clear where the configuration object is delivered to `calliope`, so that the pydantic model that #704 (and #717) use has a single place of delivery. This would be the output of the new `preprocess/model_definition.py` file.

**Comment by irm-codebase** [comment]
Documentation has been updated. The YAML section now also explains templating (since it's now a generic feature).
I've also separated the section into 'standard YAML' and our 'rich YAML', for clarity.

File: tests/test_preprocess_model_definition.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                    template: T1
+                    foo: bar
+            a:
+                template: T2
+            """
+        )
+        with pytest.raises(ModelError, match="Circular template reference detected"):
+            TemplateSolver(text)
+
+    def test_incorrect_template_placement_error(self):
```
Comment (brynpickering) [inline tests/test_preprocess_model_definition.py:203]: should we be confirming that `template` is not defined in `config`, or is that not making its way into the template solver?

Another place where `template` might now be ending up is in top-level params (the `parameters` key). We should ensure that this is working as expected.

**Review by brynpickering — CHANGES_REQUESTED** [review]
Nice. I'm on board with this. We're slowly but surely extracting content from AttrDict and will eventually be able to remove it entirely! We'll need to work out why this change has impacted the national-scale example model, though. Something about the order of dict unions is likely messing up the national-scale model.

File: src/calliope/preprocess/model_definition.py [hunk 2]
```diff
@@ -153,3 +183,80 @@ def _log_overrides(init_model_def: AttrDict, overriden_model_def: AttrDict) -> N
...
+        result.union(local_data, allow_override=True)
+
+        stack.remove(name)
+        return result
+
+    def _resolve_data(self, section, level: int = 0):
+        if isinstance(section, dict):
+            if self.TEMPLATES_SECTION in section:
+                if level != 0:
+                    raise exceptions.ModelError(
```
[the remark this answers is not in the document]
Comment (irm-codebase) [reply src/calliope/preprocess/model_definition.py:245]: The idea is to have the schema fit what the model data should have _at the moment of inclusion_. This is why I removed the templates portion of the schema. For the future `pydantic` update, this would mean that templates are not part of the validation model, simplifying our case.

File: tests/test_core_attrdict.py [hunk 3]
```diff
@@ -300,66 +207,6 @@ def test_del_key_nested(self, attr_dict):
...
-
-            assert "some_int: 10" in result
-            assert "some_float: 0.5" in result
-            assert "a_list:\n- 0\n- 1\n- 2" in result
-
-    def test_to_yaml_string(self, yaml_filepath):
-        d = AttrDict.from_yaml(yaml_filepath)
-        result = d.to_yaml()
+    def test_to_yaml_string(self, attr_dict):
+        result = attr_dict.to_yaml()
```
[the remark this answers is not in the document]
Comment (irm-codebase) [reply tests/test_core_attrdict.py:211]: I decided against it because this functionality is unique to `attrdict`, not our 'rich' YAML definition...
But I guess it makes sense to move it.

File: docs/creating/yaml.md [hunk 4]
```diff
@@ -125,9 +206,160 @@ scenarios:
     * The imported files may include further files, so arbitrary degrees of nested configurations are possible.
     * The `import` statement can either give an absolute path or a path relative to the importing file.
 
-## Overriding one file with another
+### Reusing definitions through templates
```
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/creating/yaml.md:209]: not needed, they can be used anywhere now
The code in nodes, techs and data_tables that handled this was removed in this PR.

File: docs/migrating.md [hunk 5]
```diff
@@ -181,10 +181,9 @@ This split means you can change configuration options on-the-fly if you are work
 `locations` (abbreviated to `locs` in the Calliope data dimensions) has been renamed to `nodes` (no abbreviation).
 This allows us to not require an abbreviation and is a disambiguation from the [pandas.DataFrame.loc][] and [xarray.DataArray.loc][] methods.
 
-### `parent` → `base_tech` + `template`
+### `parent` → `base_tech`
```
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/migrating.md:184]: The problem here is that this PR removes all the 'inheritance' stuff. The original text would be confusing...
Would this make sense?

---------------

```text
### `parent` → `base_tech` + `template`### `parent` → `base_tech`

Technology inheritance has been unlinked from its abstract "base" technology. Technologies must now inherit from `base_tech` which fixed to be one of [`demand`, `supply`, `conversion`, `transmission`, `storage`].
```

File: src/calliope/preprocess/model_definition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply src/calliope/preprocess/model_definition.py:245]: More explanation: if the templates section reaches the moment where the schema is applied, it means that dictionary was not produced by our model_definition function. Thus, we should not trust it and fail.

File: tests/test_preprocess_model_definition.py (hunk 1 above)
Comment (irm-codebase) [reply tests/test_preprocess_model_definition.py:203]: Both of these are fine. We should see template as something a user decides they want to re-use.
Although using it in `config` and `parameters` would be weird, it's ok as long as the final shape fits our expected schema.

This is the advantage of moving this to the IO: our job is to ensure that what is fed into calliope is good, not to police how users should / should not define their files.

Additionally, once we move the parameter definition from the schema to the model YAML, templates _would_ be useful in that section.

Comment (irm-codebase) [reply tests/test_preprocess_model_definition.py:203]: I will add this check to the 'overrides' section, though. Since it can alter templates due to our ordering (overrides > templates).
An override should not make use of `template:`, otherwise it would result in weird behaviour.

File: src/calliope/preprocess/model_definition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply src/calliope/preprocess/model_definition.py:245]: Given the developments in other PRs, the schema / `pydantic` would eventually catch this.
I added this because it feels like a useful error message, and it is within the 'scope' of the template functionality.

**Comment by irm-codebase** [comment]
Ready for review again @brynpickering! 

I've included all your suggestions, except a couple that I believe are not necessary (see the unresolved comments).

The random error in the national example model has been fixed. I must be honest in saying that I still see it as a bug, but it is not within the scope of this PR.

File: docs/migrating.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (brynpickering) [reply docs/migrating.md:184]: OK, but for migrating from having an inheritance chain (previously defined by `parent` alone) is having `base_tech` + `template`. Maybe there needs to be a separate section for `tech_groups` -> `templates`. Basically, if a user has `tech_groups`, this page should provide them with a clear pointer for how to migrate. Currently there isn't one. That `templates` can do even more than `tech_groups` is besides the point.

**Comment by brynpickering** [comment]
Thanks @irm-codebase  - I've responded. Where has the loss in coverage come from?

File: docs/creating/yaml.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/creating/yaml.md:209]: Correct. Your first example would work well.

I saw templates as purely 'extended' yaml. As in, they would work with any YAML file, and the templates code would live inside `io.py`.

You are also correct in that using templates within the override would introduce issues. My code currently does not cover for it. This is the kind of edge-case that is avoided by just solving it _before_ overrides.

So, two choices:
- Do we keep the current behaviour, and add code for the edge-case you described?
- Do we move templates to the `io.py` function (meaning that `read_rich_yaml` always solves them)?

I prefer the second even if we lose some templating power because it is more robust overall, code-wise.

**Comment by irm-codebase** [comment]
> Thanks @irm-codebase - I've responded. Where has the loss in coverage come from?

No prob. Looking at the codeconv report, it seems like it comes from the thinning of `tools.py` and other files.
This PR removed a lot of code from `src/`, so the percentages lie a bit here.

I can add tests, but they would not be for the functionality impacted by this PR.

File: docs/migrating.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/migrating.md:184]: Sounds reasonable. I will include that once other aspects of this PR are clarified.

**Comment by brynpickering** [comment]
> I can add tests, but they would not be for the functionality impacted by this PR.

Makes sense. Nah, I won't force that on you!

**Comment by irm-codebase** [comment]
An exception seems to be in `_load_yaml`... for something that might not even be used anymore.
I'll give it a look in the next round of the review.

File: docs/creating/yaml.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/creating/yaml.md:209]: Correct. I will add a function that tests for that then.
It is important because after this goes in, schemas (`pydantic`) and so on must check the data structure of the dictionary returned by `model_definition`, so this edge case will be missed.

File: src/calliope/io.py [hunk 6]
```diff
@@ -208,5 +216,116 @@ def save_csv(
...
+
+    Saves the AttrDict to the ``path`` as a YAML file or returns a YAML string
+    if ``path`` is None.
+    """
+    result = AttrDict(data).copy()
+    # Prepare YAML parsing settings
+    yaml_ = ruamel_yaml.YAML()
+    yaml_.indent = YAML_INDENT
+    yaml_.block_seq_indent = YAML_BLOCK_SEQUENCE_INDENT
+    yaml_.sort_base_mapping_type_on_output = (
```
Comment (brynpickering) [inline src/calliope/io.py:306]: This is necessary for the same reason you found with `data_tables` ordering - we want the dict ordering to remain fixed on saving to file, so we cannot have it sorted (alphabetically) by `ruamel.yaml`

**Review by brynpickering — COMMENTED** [review]
Sorry, forgot to comment on this!

File: src/calliope/io.py (hunk 6 above)
Comment (irm-codebase) [reply src/calliope/io.py:306]: I'll add a comment specifying this, then.

File: docs/creating/yaml.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (irm-codebase) [reply docs/creating/yaml.md:209]: After going through this edge-case, it seems like a non-issue. The templates algorithm will just see the 'new' `templates` or `template` requested if you change it on the data, and solve accordingly.

This is more a question of what would feel 'natural' to users: updating what `templates` 'do' within the overrides, or just seeing them as 'copy-paste' within the overrides. Either case should be basically the same in terms of the amount of bugs.

Updating the templates within the overrides seems too convoluted for me, so I would not use it, but I think it's fine to keep the code as-is.

**Review by brynpickering — APPROVED** [review]
